src/compiler/transformers/es2015.ts

```diff
@@ -1357,6 +1401,73 @@ namespace ts {
             statements.push(captureThisStatement);
         }
 
+        function prependCaptureNewTargetIfNeeded(statements: Statement[], node: FunctionLikeDeclaration, copyOnWrite: boolean): Statement[] {
+            if (hierarchyFacts & HierarchyFacts.NewTarget) {
+                let newTarget: Expression;
+                switch (node.kind) {
+                    case SyntaxKind.ArrowFunction:
+                        return statements;
+
+                    case SyntaxKind.MethodDeclaration:
+                    case SyntaxKind.GetAccessor:
+                    case SyntaxKind.SetAccessor:
+                        // Methods and accessors cannot be constructors, so 'new.target' will
+                        // always return 'undefined'.
+                        newTarget = createVoidZero();
+                        break;
+
+                    case SyntaxKind.Constructor:
+                        // Class constructors can only be called with `new`, so `this.constructor`
+                        // should be relatively safe to use.
+                        newTarget = createPropertyAccess(
+                            setEmitFlags(createThis(), EmitFlags.NoSubstitution),
+                            "constructor"
+                        );
+                        break;
+
+                    case SyntaxKind.FunctionDeclaration:
+                    case SyntaxKind.FunctionExpression:
+                        // Functions can be called or constructed, and may have a `this` due to
+                        // being a member or when calling an imported function via `other_1.f()`.
+                        newTarget = createConditional(
+                            createLogicalAnd(
+                                setEmitFlags(createThis(), EmitFlags.NoSubstitution),
+                                createBinary(
+                                    setEmitFlags(createThis(), EmitFlags.NoSubstitution),
+                                    SyntaxKind.InstanceOfKeyword,
+                                    getLocalName(node)
+                                )
+                            ),
+                            createPropertyAccess(
+                                setEmitFlags(createThis(), EmitFlags.NoSubstitution),
+                                "constructor"
+                            ),
+                            createVoidZero()
+                        );
+                        break;
+                }
+
+                const captureNewTargetStatement = createVariableStatement(

```

Comment:
`newTarget` could be `undefined` if you're somehow not handling a case - consider adding a

```ts
Debug.assert(!!newTarget, "'newTarget' should be defined for " + (ts.SyntaxKind as any)[node.kind])
```

or something similar.